### ISSUE TITLE
Transaction Signing Delay

### ISSUE THREAD

**Issue by sparrowDom** [OriginProtocol/origin#3229]
Android Coinbase Wallet, Mainnet.

When onboarding trying to publish a profile the DApp prompts me to confirm transaction in my Coinbase Wallet but the actual Coinbase prompt for signing shows up 30 seconds late. As a normal user I would probably assume something is wrong and quit (since the dapp communicates I should see the prompt immediately) 

Do not know if this has anything to do with relayer or meta-transactions...

![Screenshot_20190906-003353](https://user-images.githubusercontent.com/579910/64388477-85557c80-d03f-11e9-9f78-d9058fdfdb17.jpg)

**Comment by micahalcorn** [comment]
@sparrowDom I experienced the same in basically all of the mobile DApp browsers. ☹️ I had similar experiences when conducting actual transactions, so I suspect that this is unrelated to identity.

**Comment by micahalcorn** [comment]
<img width="868" alt="Screen Shot 2019-09-07 at 6 41 18 AM" src="https://user-images.githubusercontent.com/273937/64475678-8919f980-d13a-11e9-9d2b-7a7853993b6a.png">

**Comment by micahalcorn** [comment]
I actually just saw this in our app and I got tired of waiting for the confirmation prompt to appear. 😫 I swiped backward to the previous screen, tried again, and couldn't reproduce the seemingly-endless spinner.

**Comment by tomlinton** [comment]
There is a preflight request to the relayer to see if it is going to accept the transaction, so there is at least a few round trips that have to happen before the request gets to the signing phase. Maybe we could remove that preflight request.

**Comment by sparrowDom** [comment]
If a user is using a relayer and we are supplying the gas for the transaction execution would it make sense that we just communicate that part?

Some kind of a modal saying: `Negotiating with Origin to sponsor the gas for this transaction.` After that we could briefly display we have approved the gas for the transaction. (though that might be overwhelming to the user)

With the idea to communicate what is going on, why the user is waiting. If we are sponsoring the gas, he might not have problems waiting a bit more. In the end maybe a user could also feel rewarded since Origin has given him the gas.

**Comment by micahalcorn** [comment]
Yeah @sparrowDom I would prefer to prevent the delay, but if there's going to be one, we should be more accurate about what to expect. 😕

**Comment by mikeshultz** [comment]
This is probably related to the relayer getting clogged up.  The behavior of the dapp and relayer together could probably use a little massaging here.  Here's what currently happens when the relayer is overloaded and the dapp tries to send a tx:

1) dapp does preflight
2) Relayer does some fraud checking, returns a thumbs up
3) dapp sends final request for tx
4) Relayer essentially twiddles its thumbs waiting for a sending account to become availble
5) Dapp waits on relayer to return a tx hash and the http connection probably times out
6) Dapp does ___?

Essentially a stalemate.  the relayer may actually end up sending the transaction and returning a tx hash to no one(but in the case of ~12 yesterday, that never happened because it was restarted.  those transactions evaporated.

I'm thinking this whole "twiddle thumbs waiting for an account" was a bad design choice.  It was meant to deal with short term delays(seconds), not wait longer than an HTTP timeout.

I think the account acquisition needs to timeout after say, 5s and return an error to the client instead of just timing out with no resolution.  

The dapp also should maybe handle the case of not getting a tx hash back from relayer(for whatever reason) and consider it a failure.  Anything is better than an endless spinner.

**Comment by sparrowDom** [comment]
I like both of your suggestions @mikeshultz . If I understood correctly: 
- the relayer would return an `can not acquire account` error after ~5 seconds and the Dapp would  display an easy understandable error message to the user
- dapp handling the timeouts itself

Though I think this is not related to the endless spinner thing (I might be wrong though). But I think the spinner only displays on the Dapp once the transaction hash has been received. Before the hash is received it displays the prompt to sign the transaction - even though the wallet hasn't received anything to sign yet. I think there should be a pre-state to the "sign the transaction prompt" that should communicate there are negotiations with relayer going on. And that part should be timeoutable too.

**Comment by mikeshultz** [comment]
I thought there was a (confusing) transaction spinner also before the purchase confirm where the dapp communicated with the relayer?  If not, no problem, I'm in the wrong issue here and will open up another.

**Comment by mikeshultz** [comment]
Added @sparrowDom as my co-assignee on this one.  I still think there's a chance all my junk above is a separate issue.

**Comment by sparrowDom** [comment]
@mikeshultz sorry for dropping the ball on this one. Thanks for the above debug info I think that sheds some light on the possible delays. From what I understand this is (at least) an UX problem. 

**The current state:**
From UX point of view when a user initiates a transaction 2 different modals are displayed to the user: 
- Modal 1 waiting for transaction signature/confirmation saying:
  - *Desktop* Please confirm this transaction in MetaMask
  - *Coinbase Wallet* Please confirm this transaction in Coinbase Wallet
  - *Origin Marketplace* Waiting for Confirmation
- Then the transaction is confirmed either by user or our backend
- Modal 2: Writing to the blockchain. This might take a minute

What I think the problem is (like you point out @mikeshultz) that there is a lot of things happening in the period of Modal 1 being active that can result in variable delays (larger than 15 seconds in some cases) before user is able to perform any actions.

My biggest issue here is that we communicate to the user to take an action of confirming the transaction when in reality the Wallet hasn't received the prompt for signing the transaction. And because of the wording in **Modal 1** this confusion is the biggest in third party mobile wallets and Desktop.

We could do any of the following:
1. Change the wording in Modal 1 for non Origin Marketplace App transactions to something more along the lines of: "Making preparations to sign the transaction" (Very easy to implement)
2. Instead of having 2 Modals for the whole flow introduce the 3rd modal that shows up immediately when the pre-flight flow starts. As in point 1. the modal would say "Making preparations to sign the transaction". Once the wallet receives the prompt to sign the transaction the modal we see now on production would appear saying. "Please confirm this transaction in MetaMask/Coinbase Wallet"
3. Investigate if delays are fixable on the backend. 

I vote to research how involved is fix **nr. 2**. If it is more involved just do fix **nr. 1** and schedule fix **nr. 2** with lower priority. I would leave it up to @mikeshultz to decide what to do about **nr. 3** since he is better equipped to make a decision. 

Thoughts? @micahalcorn @mikeshultz

**Comment by mikeshultz** [comment]
I def think option 1 should be done.  It is confusing and a simple change.

I was poking around the dapp last week and I couldn't really figure out what was getting done during that first modal, however(it looked pure UI to me).  All of those network requests I profiled above occur during the second modal, I think.  Even the relayer preflight request was basically right before sign+send. Please correct me if I'm wrong, I still get lost in the dapps sometimes :confused:.

**Comment by sparrowDom** [comment]
Aha @mikeshultz thanks. Let me debug what is going on when the 1st modal is active, and lets have  @micahalcorn weigh in on the UX decision.

**Comment by mikeshultz** [comment]
Some of the issues from #3410 may be related.

**Comment by sparrowDom** [comment]
@mikeshultz have done some testing with updating the profile. And while the first modal is active this code gets called: 
[deployIdentity](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/mutations/_txHelper.js#L143-L225)  and the this one [txHelper](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/mutations/_txHelper.js#L143-L225) and the relayer one [relayer](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/mutations/_relayer.js#L9-L75) and only [on this line](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/mutations/_relayer.js#L76) is when the actual signature / user input is requested.  So there are a bunch of non instant calls being made even [call to relayer](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/mutations/_relayer.js#L27) that are executed before the user is able to sign a transaction. 

I think it would be cool that before the signature happens in the above mentioned code pieces the 3rd modal ("preparing transaction") should be presented to the user.

**Comment by mikeshultz** [comment]
ok, so it runs through building the transaction (before signing) in that first modal.  That's ~90% of my network request profiling above...  And any of the `Failed to fetch`(#3410) errors would potentially have some unintended side-effects.  I think it might be beneficial to run through this stuff, add defensive checks, and bail early if there's any issues.  Kind of wanting to do a bit of a refactor to `txHelper` as well.

I'm not sure, but maybe there's a better way to deal with error handling in the dap as well.  For instance, it doesn't looks like `Buy.onClick()` has any error handling if something goes wrong in the `makeOffer()` callback.  But I defer to @sparrowDom on that.

**Comment by micahalcorn** [comment]
In my opinion...

1. Don't show a spinner when the user is supposed to take an action.
2. Don't tell the user to take an action if we're still waiting for the relayer.
3. Prevent the delay with the relayer (or whatever the culprit it).

If 3️⃣ is achieved, then I don't think we would want to end up with a very short-lived modal state that flickers and can't be read.

**Comment by mikeshultz** [comment]
@tomlinton noticed that during this EventCache is pulling event logs.  Unfortunate timing of something that used to happen earlier during the app launch.  With PR #3430, it has been reduced, but there's still some queries to eventCache that are causing this to continue to happen(in an apparently reduced fashion).  This should be somewhat improved, but I will be trying to ferret these other queries out today.

**Comment by franckc** [comment]
Thanks @mikeshultz for getting to the bottom of this - I think this could very well explain those long delays. Once we have fixed all those queries, I'd be in favor of completely disabling EventCache on the client when it's running in performance mode. Otherwise I fear in the future someone may had a query that will again cause EventCache to get re-activated. This can be done in a separate PR.

**Comment by mikeshultz** [comment]
That seems extreme. I like the idea of having it there as a fallback, should the server be down or something.  I think everyone just needs to be mindful of the impact of how some of these queries are written.

**Comment by mikeshultz** [comment]
I think we might be able to close this one after the identity query improvements that triggered EventCache's rapid fire fetching.  I think there's still room for improvement to be made here, but I think the original purpose of this issue has been resolved.

**Comment by mikeshultz** [comment]
I'm gonna close this.  There's a ton here that can still be addressed(UI of relayer fallback, performance, error handling), but I think the original reason for submission has been taken care of.  If anyone disagrees, please reopen.

**Comment by micahalcorn** [comment]
@sparrowDom in the mobile wallet apps, we're not able to detect when the signature prompt is actually displayed to the user, are we? We only know that we have asked the wallet to request a signature and then we eventually see that the user clicked one of the two buttons? 🖱

**Comment by sparrowDom** [comment]
@micahalcorn yup exactly.

Though the time between when we ask for the signature to be displayed and when it actually is - is according to my testing very small. The delay that was initially reported in this ticket was due to our client side code doing stuff before handing the signature off to the wallet
